### DISCUSSION TITLE
zh/guides/sahi-tiled-inference/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#16789] [Docs]
# zh/guides/sahi-tiled-inference/

Learn how to implement YOLO11 with SAHI for sliced inference. Optimize memory usage and enhance detection accuracy for large-scale applications.

https://docs.ultralytics.com/zh/guides/sahi-tiled-inference/

**Comment by bqpeng** [comment]
Hi, inquiry. is SAHI only used in the reasoning stage? For example, if I have a customized dataset and train it, I still train it normally and get the pt file. Then add SAHI and use the trained pt file for inference, still for training, I need to add SAHI as well.

**Reply by glenn-jocher** [reply]
@bqpeng sAHI is used only during the inference stage. You train your model normally and then apply SAHI for inference to handle large images efficiently. For more details, please refer to the [SAHI guide](https://docs.ultralytics.com/guides/sahi-tiled-inference/).

**Reply by bqpeng** [reply]
thank you！

**Reply by glenn-jocher** [reply]
@bqpeng thank you for your interest! SAHI is used only during the inference stage to optimize detection on large images. You can train your model as usual and then apply SAHI for efficient inference. For more details, please refer to the SAHI guide.

**Reply by mmoollllee** [reply]
However, you need to make sure that you are training square images that actually represent what SAHI will use for inference, right? Check out my [tiling plugin for Voxel51](https://github.com/mmoollllee/fiftyone-tile) to tile your hq images for training.

**Reply by glenn-jocher** [reply]
Thank you for sharing! Yes, ensuring that your training images match the tiled inference setup used by SAHI is crucial for consistent results. Your Voxel51 tiling plugin looks like a valuable tool for preparing datasets for such workflows. For more on integrating SAHI with YOLO11, please refer to the [official SAHI guide](https://docs.ultralytics.com/guides/sahi-tiled-inference/).

**Comment by UltralyticsAssistant** [comment]
👋 Hello, thank you for your interest in Ultralytics 🚀! We recommend checking out our comprehensive [Documentation](https://docs.ultralytics.com) for guidance and insights. You'll find examples on using [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) tools there, which may address many of your questions.

If you're encountering any issues with the SAHI tiled inference guide, please make sure to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) if this is a 🐛 Bug Report, to help us understand and address it effectively.

For detailed questions or custom training inquiries ❓, include comprehensive information such as dataset image samples and training logs. Make sure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the vibrant Ultralytics community where you feel most comfortable. Engage in real-time discussions on [Discord](https://ultralytics.com/discord) 🎧, explore deep-dive topics on [Discourse](https://community.ultralytics.com), or share insights on our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure you have the latest version of the `ultralytics` package installed, along with all necessary [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a compatible [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be operated in a variety of verified environments, complete with necessary dependencies like [CUDA](https://developer.nvidia.com/cuda) and [CUDNN](https://developer.nvidia.com/cudnn):

- **Notebooks** with free GPU access: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM – [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI – [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image** – [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

Check if our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are green. This badge verifies operation accuracy for all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across several platforms daily and on each commit:

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

An Ultralytics engineer will also assist you soon 😊.

**Comment by ZSBSB** [comment]
Your work is perfect, I have one question, does the current point SAHI library support yolov11's rotating frame target detection?

**Reply by fcakyon** [reply]
Yes it supports 👌🏻

**Comment by handsome2456** [comment]
Hi! How to utilize sahi for YOLOv8 instance segmentation?

**Reply by glenn-jocher** [reply]
@handsome2456 to utilize SAHI with YOLOv8 for instance segmentation, you can integrate them by installing the required libraries (`pip install -U ultralytics sahi`), loading a YOLOv8 model with `AutoDetectionModel` from SAHI, and performing sliced inference using `get_sliced_prediction`. You can find detailed implementation steps in the [SAHI Tiled Inference guide](https://docs.ultralytics.com/guides/sahi-tiled-inference/). Let me know if you need further assistance!

**Comment by bearllm** [comment]
Does SAHI support Oriented Bounding Box?

**Reply by glenn-jocher** [reply]
@fcakyon could you answer @bearllm on SAHI support for YOLO11 OBB models?

**Reply by fcakyon** [reply]
Hello @bearllm, author of the SAHI here. 

SAHI supports YOLO11 OBB models as of `sahi==0.11.20` 🚀

**Comment by zhugelaozei** [comment]
Hi，inquriy.How to perform inference on multiple slices simultaneously?I need to infer a so high resolution image and I want to maintain inference speed, but it is so slow because batch size and cuda is only used one.

**Reply by glenn-jocher** [reply]
@zhugelaozei to perform inference on multiple slices simultaneously and optimize the use of CUDA resources, you can use SAHI's batch prediction capabilities. This allows you to process multiple slices or images in parallel by leveraging batch size and slicing configurations. Refer to the SAHI documentation for batch prediction examples: https://docs.ultralytics.com/zh/guides/sahi-tiled-inference/. Ensure your hardware and CUDA setup are configured to handle parallel processing efficiently.

**Comment by wenxi2** [comment]
hi，is there code for evaluating the COCO metrics after cutting and stitching predictions?

**Reply by fcakyon** [reply]
@wenxi2 you can follow this guide for SAHI based YOLOv8/YOLO11/RTDETR evalution: https://github.com/obss/sahi/discussions/622

**Comment by wenxi2** [comment]
Thank you very much for your reply. I followed the tutorial for the evaluation and obtained such results, but the visualization effect has clearly improved.I am using the VisDrone dataset.&nbsp;Do you know the reason?






随便
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ultralytics/ultralytics"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年12月16日(星期一) 晚上9:40
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ultralytics/ultralytics] zh/guides/sahi-tiled-inference/ (Discussion #16789)





 
@wenxi2 you can follow this guide for SAHI based YOLOv8/YOLO11/RTDETR evalution: obss/sahi#622

**Reply by fcakyon** [reply]
Please try decreasing the threshold to 0.001 to increase the COCO eval scores @wenxi2 .

**Comment by echo385** [comment]
hi, when I use the sari to infer, how can I use two gpus to infer my images?

**Reply by glenn-jocher** [reply]
@echo385 to use multiple GPUs with SAHI for inference, you can implement custom parallel processing by splitting your workload across devices. Since SAHI doesn't natively support multi-GPU inference in its current implementation, you'll need to manually distribute batches between GPUs using PyTorch's parallelization tools like `DataParallel` or process images separately on each device. For detailed guidance on GPU utilization patterns, refer to the SAHI GitHub repository's discussions and issues.

**Comment by Ares-cz** [comment]
Hi, inquiry.
I checked SAHI's repository and it seems that the evaluation script they built is based on the COCO format (https://github.com/obss/sahi/blob/main/docs/cli.md#coco-evaluate-command-usage), but when building the YOLO dataset, the dataset format is based on /images and /labels. Does this mean that I need to convert the YOLO dataset into the COCO format to meet the verification script? But is there a script to convert YOLO to COCO format?

**Reply by glenn-jocher** [reply]
@Ares-cz for using SAHI's COCO evaluation with YOLO-formatted datasets, you'll need to convert your labels. While Ultralytics doesn't provide a dedicated YOLO-to-COCO converter, you can use Python scripts with `pycocotools` or third-party converters like https://github.com/Taeyoung96/Yolo-to-COCO-format-converter. For SAHI integration, consider using their `sahi.utils.coco` tools to convert predictions to COCO format post-inference.

**Reply by fcakyon** [reply]
@Ares-cz you can use yolo to coco conversion utitilites available inside fiftyone package

I'm the creator and maintainer of sahi, feel free to ask more questions by tagging me.

**Comment by Tony2408** [comment]
Hi. I've applied sahi and yolov11 to infer the bbox of small instance. What confuses me is that multiple local regions of some large targets are individually detected as bounding boxes. I would like to ask: Can this situation only be mitigated by adjusting the slice size and confidence threshold? Are there parameters similar to global NMS that can be adjusted?

**Reply by fcakyon** [reply]
Hey, SAHI author here 👋🏻 

@Tony2408, it means your slices are smaller than your objects. You can slightly increase the slice height and width or you apply [Non-Maximum-Merging](https://github.com/obss/sahi/blob/4f427d4e8dcb2347b01eff521a6473e603ba61c5/sahi/predict.py#L156) to combine these predictions that belong to same object.

**Reply by Tony2408** [reply]
Great work! Thank you for your reply. Increasing the height and width of the slices for the same image has significantly improved the results. However, I'm wondering if this means that when my task requires identifying multi-scale targets in an image, it's difficult for Sahi's inference method to cover all scales?

**Reply by glenn-jocher** [reply]
@Tony2408 You've identified an important challenge with tiled inference approaches. When dealing with multi-scale targets, there's indeed a trade-off: larger slices work better for big objects but might be less efficient for small ones, while smaller slices excel at detecting tiny objects but fragment larger ones.

Some approaches to handle multi-scale targets with SAHI:

1. Use multiple inference passes with different slice configurations and merge results
2. Adjust overlap ratios (higher overlap can help with object fragmentation)
3. Customize the Non-Maximum-Merging parameters to better combine fragmented detections

The SAHI library continues to evolve with improvements to handle these scenarios. For particularly challenging multi-scale detection tasks, you might also consider combining SAHI with model ensembling techniques for more robust results.

**Comment by YWJ-sudo** [comment]
Hello author, I am encountering an issue when using SAHI for sliced prediction. When I change the model_path in detection_model to the path of my locally trained weight file (best.pt), it produces an error. However, if I rename best.pt to yolov8n.pt, it can run successfully.What's going on here? How should I quickly use my own weight file?

ERROR：TypeError: ('model_path is not a valid Ultralytics model path: ', SSLError(MaxRetryError('HTTPSConnectionPool(host=\'api.github.com\', port=443): Max retries exceeded with url: /repos/ultralytics/assets/releases/tags/v8.3.0 (Caused by SSLError("Can\'t connect to HTTPS URL because the SSL module is not available."))')))

**Reply by YWJ-sudo** [reply]
Sorry, it's been resolved now. There were just some minor errors.

**Reply by fcakyon** [reply]
@YWJ-sudo I'm glad to hear it's resolved. Feel free to tag me if you have any more SAHI related issues (SAHI author and core maintainer here) 👍🏻

**Reply by YWJ-sudo** [reply]
How to obtain the COCO format annotation file  after batch patch prediction?

**Reply by YWJ-sudo** [reply]
如果我有一张1600*1600的图片，在预处理阶段我先将其切割成4个切片，即每个切片尺寸为400*400，并以此训练模型，而当预测时，我在原图上进行预测，且使用sahi切片推理并将切片高宽设置为400，这会不会得到更好的效果

**Reply by fcakyon** [reply]
@YWJ-sudo when you provide `coco_dataset_json` to sahi predict cli or function, it saves results in coco prediction format: [cli usage](https://github.com/obss/sahi/discussions/622) | [functional usage](https://github.com/obss/sahi/blob/e7109ee3a117d421f64ed383ad4a073b8ff7b000/sahi/predict.py#L411)

Or you can convert each ObjectPrediction object to COCO prediction format by using `object_prediction.to_coco_prediction(image_id=...)`: https://github.com/obss/sahi/blob/e7109ee3a117d421f64ed383ad4a073b8ff7b000/sahi/prediction.py#L122

**Reply by fcakyon** [reply]
@YWJ-sudo based on your second question. If you are cropping images to 400x400, then resize to 640 on training (ultralytics does this by default), then you should use 400 as slice height and width in `sahi`, and that's all (sahi ultralytics integration resizes crops to 640 before inference by default).

Optionally (I don't suggest this unless you know what you are doing), if you want `sahi` to crop images to 400, then perform inference without resizing the crops, then you should set [image_size in UltralyticsDetectionModel/AutoDetectionModel class as 400](https://github.com/obss/sahi/blob/e7109ee3a117d421f64ed383ad4a073b8ff7b000/sahi/auto_model.py#L33).

**Comment by YWJ-sudo** [comment]
你好，请问预测时出现预测的锚框的长或宽为0的无效锚框，那么其是否会自动过滤这些无效锚框

**Reply by glenn-jocher** [reply]
@YWJ-sudo 一般情况下 Ultralytics 的后处理不会返回宽或高为 0 的检测框；若在 SAHI 切片拼接或导出为 COCO 后仍出现，可在评估前用 SAHI 的 `remove_invalid_coco_results` 进行清理，示例：
```python
from sahi.utils.coco import remove_invalid_coco_results
from sahi.utils.file import save_json

coco_results = remove_invalid_coco_results("coco_results.json")  # 过滤无效/越界框
save_json(coco_results, "coco_results_fixed.json")
```
函数示例见[SAHI COCO utilities 文档](https://obss.github.io/sahi/coco/)。

**Comment by Mz-Han** [comment]
Hello. When using YOLO26 for instance segmentation, I encountered an issue of insufficient CUDA memory and plan to solve it through slice inference. The current environment is torch==1.13.1+cu116, torchvision==0.14.1+cu116, torchaudio==0.13.1, while installing the latest version of Sahi requires PyTorch 2.4+ as a prerequisite. Is there a version of Sahi that can be installed without upgrading PyTorch? It would be greatly appreciated if you could provide an example of inference visualization. I have installed sahi==0.11.15, but this version cannot output masks

**Reply by glenn-jocher** [reply]
@Mz-Han `sahi==0.11.15` is too old for Ultralytics segmentation mask support, so getting `mask=None` there is expected; mask output was added in newer SAHI releases (see the SAHI maintainer note in [obss/sahi discussion #1000](https://github.com/obss/sahi/discussions/1000)) and the current [SAHI PyPI page](https://pypi.org/project/sahi/) shows the latest version stream. If you’re pinned to `torch==1.13.1`, you can install/upgrade SAHI without forcing a PyTorch upgrade by avoiding dependency resolution, but to actually run YOLO26-seg + sliced masks you’ll likely need a newer PyTorch/Ultralytics stack—please paste `!yolo checks` (or `ultralytics.utils.checks.collect_system_info()`) and your full install command here so we can confirm what’s constraining your environment; also restart your notebook kernel after any upgrades. For visualization (without SAHI), this works for YOLO26 segmentation:

```python
from ultralytics import YOLO

model = YOLO("yolo26n-seg.pt")
results = model.predict(source="path/to/image.jpg", imgsz=640, conf=0.25, device=0)
results[0].plot(pil=True).save("pred.png")
```

If you do proceed with SAHI, the parameter/visual export pattern is the same as in the [SAHI sliced inference guide](https://docs.ultralytics.com/zh/guides/sahi-tiled-inference/), but you’ll need a newer SAHI version than `0.11.15` to get masks.

**Reply by Mz-Han** [reply]
Hello, thank you for your reply. I have upgraded PyTorch. How can I display only the mask? My image resolution is 3807*3000, with over a thousand small targets. How can I improve the running speed

**Reply by glenn-jocher** [reply]
@Mz-Han To visualize/save *only* the mask, render the pixel-space polygons from `results[0].masks.xy` (see the `Masks` outputs in [Predict mode docs](https://docs.ultralytics.com/modes/predict/)) into a blank image, e.g.:

```python
from ultralytics import YOLO
import cv2
import numpy as np

model = YOLO("yolo26n-seg.pt")
img_path = "path/to/image.jpg"

img = cv2.imread(img_path)
h, w = img.shape[:2]

r = model.predict(source=img_path, device=0, imgsz=1280, conf=0.25)[0]
mask_img = np.zeros((h, w), dtype=np.uint8)

if r.masks is not None:
    for seg in r.masks.xy:  # list of Nx2 polygons in pixel coords
        cv2.fillPoly(mask_img, [seg.astype(np.int32)], 255)

cv2.imwrite("mask.png", mask_img)
```

For speed on 3807×3000 with 1000+ small objects, sliced inference runtime is dominated by slice count, so increase `slice_height/slice_width` and reduce overlap until you stop missing small targets (and if you’re on NVIDIA, exporting to TensorRT can help; if you share `!yolo checks` plus your current slice settings we can suggest concrete values based on your GPU/CPU).
